### ISSUE TITLE
Guards in patterns don't narrow types

### ISSUE THREAD

**Issue by rshin** [microsoft/pyright#4434]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Given a case in pattern matching like `case _ if isinstance(x, TypeName):`, pyright doesn't realize that in subsequent cases, `x` cannot be of type `TypeName`.

**To Reproduce**
Consider the following code:

```python
from typing import TypeAlias
from typing_extensions import assert_never

class X: pass
class Y: pass
class Z: pass

XY: TypeAlias = X | Y
XYZ: TypeAlias = X | Y | Z

def f(arg: XYZ) -> None:
    match arg:
        case _ if isinstance(arg, XY):
            print("Got X or Y")
        case Z():
            print("Got Z")
        case _:
            assert_never(arg)

f(X())
f(Y())
f(Z())
```

When run, it prints
```
Got X or Y
Got X or Y
Got Z
```

Pyright incorrectly complains with `error: Argument of type "X | Y" cannot be assigned to parameter "__arg" of type "Never" in function "assert_never"`.

However, pyright correctly finds no fault with the following code:
```python
def g(arg: XYZ) -> None:
    if isinstance(arg, XY):
        print("Got X or Y")
    elif isinstance(arg, Z):
        print("Got Z")
    else:
        assert_never(arg)
```

As `case XY():` doesn't work (Python complains with `TypeError: called match pattern must be a type`), the above is a workaround since https://peps.python.org/pep-0604/#isinstance-and-issubclass is available.

**Expected behavior**
No errors reported in the above example code.

**Screenshots or Code**
See above

**VS Code extension or command-line**
pyright 1.1.287 on the command-line

**Comment by erictraut** [comment]
Yes, this is by design. Guard expressions do not apply to type narrowing. You should use `X()|Y()` if you want type narrowing to apply.

**Comment by rshin** [comment]
Thanks for the fast followup. I noticed that mypy does apply type narrowing in guard expressions, and the above code works as I expected with after replacing `isinstance(arg, XY)` with `isinstance(arg, X | Y)`. What is the reason for deciding differently with Pyright?

**Comment by erictraut** [comment]
Pyright's support for pattern matching predates that of mypy, so you could ask the mypy maintainers why they made different implementation decisions than pyright. :)

There will undoubtedly be differences between the two in certain details. I recently wrote [some documentation](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md) that you might find interesting.

In this particular case, I didn't consider it important to add type narrowing for guard expressions because it's an anti-pattern to use the subject expression in a guard when you could use the core pattern matching facility to test for it.
